File: Stack/Monotonic Stack/Lexicographically smallest after removing k/stack_app.cpp
#include<bits/stdc++.h>
using namespace std;

// Time complexity :- O(n + k)
// Space complexity :- O(n)

// Approach :- 
// we can update the value of k depending of the size of s 
// maintain a monotnic stack of char and pop the char for only 'k' times 
// remaining string will be our answer 

// Link :- https://www.geeksforgeeks.org/problems/lexicographically-smallest-after-removing-k/1

class Solution {
  private:
    bool isPowerOf2(int num){
        return (num & (num - 1)) == 0;
    }
  
  public:
    string lexicographicallySmallest(string &s, int k) {
        int n = s.size();
        
        // check if length is power of 2 or not 
        if(isPowerOf2(n)){
            k /= 2;     // divide k value by 2 
        }else{
            k *= 2;     // multiply k by 2
        }
        
        stack<char> st;
        
        for(auto ch: s){
            // maintain a montonic stack of char until we've k value 
            while(!st.empty() && k > 0 && (st.top()-'a') > (ch - 'a')){
                st.pop();
                k -= 1;
            }
            // add current char in stack 
            st.push(ch);
        }
        // remove remaining k char from the string 
        while(!st.empty() && k > 0){
            st.pop();
            k -= 1;
        }
        // check if string is empty 
        if(st.empty())         return "-1";
        
        string res = "";
        
        // form the string from remaining char 
        while(!st.empty()){
            res += st.top();
            st.pop();
        }
        // reverse the string to get answer due stack property 
        reverse(res.begin(), res.end());
        return res;
    }
};

File: Stack/Monotonic Stack/Smallest Subsequence of Distinct Characters/stack_app.java
import java.util.*;

// Time complextiy :- O(n)
// Space complexity :- O(n)

// Approach :-
// use stack to maintain a montonic stack of char which will give us lexiographically smallest string 
// we'll remove the char from the stack based on monotonicity and 
// if more occurence of curr char which we're removing from stack is present in later part of string 
// at last build the answer from the char present in the stack 


// Links :- 
// https://leetcode.com/problems/remove-duplicate-letters/description/
// https://leetcode.com/problems/smallest-subsequence-of-distinct-characters/description/

class Solution {
    public String smallestSubsequence(String s) {
        int[] freq = new int[26];

        // count the frequency of each char 
        for(char ch: s.toCharArray()){
            freq[ch-'a'] += 1;
        }

        boolean[] check = new boolean[26];
        Stack<Character> st = new Stack<>();

        // traverse on every char of string 
        for(char ch: s.toCharArray()){
            // maitain a monotonic stack while maintain the distinct char 
            while(!st.empty() && check[ch - 'a'] == false && ch < st.peek() && freq[st.peek() - 'a'] >= 1){
                check[st.peek() - 'a'] = false;
                st.pop();
            }
            // add the char into the stack only if it's not present 
            if(check[ch - 'a'] == false){
                st.push(ch);
                check[ch - 'a'] = true;
            }
            // decreament the frequency of the char which are processed 
            freq[ch - 'a'] -= 1;
        }

        StringBuilder res = new StringBuilder();

        // form the lexiographically smallest string from the char present in stack
        while(!st.empty()){
            res.append(st.peek());
            st.pop();
        }
        // reverse the string to get the right order 
        res.reverse();
        return res.toString();
    }
}